lib/absinthe/middleware.ex

```diff
@@ -0,0 +1,264 @@
+  field on that object. Concretely then, the function is called 3 times for that
+  document, with the following arguments:
+
+  ```
+  YourSchema.set_middleware(lookup_user_field_of_root_query_object, root_query_object)
+  # IO.inspect output: [{Absinthe.Resolution, #Function<20.52032458/0>}]
+  YourSchema.set_middleware(name_field_of_user, user_object)
+  # IO.inspect output: []
+  YourSchema.set_middleware(age_field_of_user, user_object)
+  # IO.inspect output: []
+  ```
+
+  In the latter two cases we see that the middleware list is empty. In the first
+  case we see one middleware, which is placed by the `resolve` macro used in the
+  `:lookup_user` field.
+
+  ### Default Middleware
+
+  One use of `set_middleware/2` is setting the default middleware on a field,
+  replacing the `default_resolver` macro. By default middleware is placed on a
+  field that looks up a field by its snake case identifier, ie `:resource_name`.
+  Here is an example of how to change the default to use a camel cased string,
+  IE, "resourceName".
+
+  ```
+  def set_middleware(%{middleware: []} = field, _object) do
+    camelized =
+      field.identifier
+      |> Atom.to_string
+      |> Macro.camelize
+
+    middleware = [{{__MODULE__, :get_camelized_key}, camelized}]
+
+    %{field | middleware: middleware}
+  end
+  def set_middleware(field, _object) do
+    field
+  end
+
+  def get_camelized_key(%{source: source} = res, key) do
+    %{res | state: :resolved, value: Map.get(source, key)}
+  end
+  ```
+
+  There's a lot going on here so let's unpack it. The first thing to note
+  is that we're using two clauses. We only want to set this middleware if there
+  is not already middleware defined (by a resolve function or otherwise), so we
+  pattern match on an empty list. Generating the camelized key is a simple matter
+  of camelizing the field identifier.
+
+  Next we need to build a list that defines what middleware we want to use. The
+  form we're using is `{{MODULE, :function_to_call}, options_of_middleware}`. For
+  our purposes we're simply going to use a function in the schema module itself
+  `get_camelized_key`.
+
+  Like all middleware functions, it takes a resolution struct, and options. The
+  options is the caemlized key we generated. We get the camelized string from
+  the parent map, and set it as the value of the resolution struct. Finally we
+  mark the resolution state `:resolved`.
+
+  ### Object Wide Authentication
+
+  Let's use our authentication middleware from earlier, and place it on every
+  field in the query object.
+
+  ```
+  defmodule MyApp.Web.Schema do
+    use Absinthe.Schema
+
+    query do
+      field :private_field, :string do
+        resolve fn _, _ ->
+          {:ok, "this can only be viewed if authenticated"}
+        end
+      end
+    end
+
+    def set_middleware(field, %Absinthe.Type.Object{identifier: :query}) do
+      field
+      |> Map.update!(:middleware, [MyApp.Web.Authentication | &1])
```

Comment:
For documentation examples, is there too much going on here language-wise or should we expect all readers of absinthe documentation to have a minimum level of familiarity with some of the fun language features of elixir like `[Module | &1]`?